components/patient-data/impl/src/test/java/org/phenotips/data/internal/controller/AbstractComplexControllerTest.java

```diff
@@ -0,0 +1,428 @@
+
+        verify(this.mocker.getMockedLogger()).error("Could not find requested document or some unforeseen error has " +
+            "occurred during controller loading ", exception.getMessage());
+        Assert.assertNull(result);
+    }
+
+    @Test
+    public void loadCatchesExceptionWhenPatientDoesNotHavePatientClass() throws ComponentLookupException
+    {
+        doReturn(null).when(this.doc).getXObject(Patient.CLASS_REFERENCE);
+
+        PatientData<String> result = this.mocker.getComponentUnderTest().load(this.patient);
+
+        verify(this.mocker.getMockedLogger()).error("Could not find requested document or some unforeseen"
+            + " error has occurred during controller loading ", PatientDataController.ERROR_MESSAGE_NO_PATIENT_CLASS);
+        Assert.assertNull(result);
+    }
+
+    @Test
+    public void loadReturnsAllData() throws ComponentLookupException
+    {
+        String datum1 = "datum1";
+        String datum2 = "datum2";
+        String datum3 = "datum3";
+        doReturn(datum1).when(this.baseProperty1).getValue();
+        doReturn(datum2).when(this.baseProperty2).getValue();
+        doReturn(datum3).when(this.baseProperty3).getValue();
+
+        PatientData<String> result = this.mocker.getComponentUnderTest().load(this.patient);
+
+        Assert.assertEquals(datum1, result.get(PROPERTY_1));
+        Assert.assertEquals(datum2, result.get(PROPERTY_2));
+        Assert.assertEquals(datum3, result.get(PROPERTY_3));
+        Assert.assertEquals(3, result.size());
+    }
+
+    @Test
+    public void loadConvertsCodeFieldsWhenControllerIsOnlyCodeFields() throws Exception
+    {
+        this.documentAccessBridge = this.codeFieldImplMocker.getInstance(DocumentAccessBridge.class);
+        doReturn(this.doc).when(this.documentAccessBridge).getDocument(this.patientDocument);
+        List<String> list1 = new LinkedList<>();
+        list1.add("HP:00000015");
+        doReturn(list1).when(this.baseProperty1).getValue();
+        doReturn(list1).when(this.baseProperty2).getValue();
+
+        PatientData<List<VocabularyProperty>> result =
+            this.codeFieldImplMocker.getComponentUnderTest().load(this.patient);
+
+        List<VocabularyProperty> propertyOneList = result.get(PROPERTY_1);
+        List<VocabularyProperty> propertyTwoList = result.get(PROPERTY_2);
+        Assert.assertNotNull(propertyOneList);
+        Assert.assertNotNull(propertyTwoList);
+    }
+
+    //-----------------------------------save() tests-----------------------------------
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void saveIsUnsupported() throws ComponentLookupException
+    {
+        this.mocker.getComponentUnderTest().save(this.patient);
+    }
+
+    //-----------------------------------writeJSON() tests-----------------------------------
+
+    @Test
+    public void writeJSONReturnsWhenGetDataReturnsNull() throws ComponentLookupException
```

Comment:
The problem is that `writeJSON()`'s return type is `void`, so it doesn't return null when `getData()` returns null; it simply returns without performing any further actions. I suppose this test could be renamed: `writeJSONChecksForDataBeforeWritingToJSON`?